File: Manipulator3D_Ruby/src/sim/trajectory.rb
# frozen_string_literal: true

require_relative "../raylib_bootstrap"
require_relative "../util/math3d"

module Sim
  class LinearTrajectory
    def initialize
      @a = Vector3.create(0.0, 0.0, 0.0)
      @b = Vector3.create(0.0, 0.0, 0.0)
      @duration = 1.0
      @t = 0.0
      @alpha = 0.0
      @finished = true
    end

    def reset(from, to, duration_sec)
      @a = Vector3.copy_from(from)
      @b = Vector3.copy_from(to)
      @duration = (duration_sec.to_f > 1e-6) ? duration_sec.to_f : 1e-6
      @t = 0.0
      @alpha = 0.0
      @finished = false
    end

    def update(dt)
      return if @finished
      @t += dt.to_f
      @alpha = Util::Math3D.clamp(@t / @duration, 0.0, 1.0)
      @finished = true if @alpha >= 1.0
    end

    def position
      Vector3.create(
        @a.x + (@b.x - @a.x) * @alpha,
        @a.y + (@b.y - @a.y) * @alpha,
        @a.z + (@b.z - @a.z) * @alpha
      )
    end

    def finished? = @finished
    def alpha = @alpha
  end
end

File: Manipulator3D_Python/src/sim/trajectory.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Tuple


Vec3 = Tuple[float, float, float]


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


@dataclass
class LinearTrajectory:
    a: Vec3 = (0.0, 0.0, 0.0)
    b: Vec3 = (0.0, 0.0, 0.0)
    duration: float = 1.0
    t: float = 0.0
    alpha: float = 0.0
    finished: bool = True

    def reset(self, from_pos: Vec3, to_pos: Vec3, duration_sec: float) -> None:
        self.a = from_pos
        self.b = to_pos
        self.duration = duration_sec if duration_sec > 1e-6 else 1e-6
        self.t = 0.0
        self.alpha = 0.0
        self.finished = False

    def update(self, dt: float) -> None:
        if self.finished:
            return
        self.t += dt
        self.alpha = _clamp(self.t / self.duration, 0.0, 1.0)
        if self.alpha >= 1.0:
            self.finished = True

    def position(self) -> Vec3:
        ax, ay, az = self.a
        bx, by, bz = self.b
        a = self.alpha
        return (ax + (bx - ax) * a, ay + (by - ay) * a, az + (bz - az) * a)
